### PR TITLE
[READY] [PORT] Making the RD Servers do something than take up space from /tg/

### PR DESCRIPTION
**Pull request by MCterra10**
## About The Pull Request
Ports the closed PR tgstation/tgstation#50774.
It's actually a _really_ neat change and it's a shame to see oranges close it because "muh design document". I wanted to get in touch with the author, but couldn't find an email or discord handle. All credit to them.

From the original PR:
```
My OCD finally couldn't take it anymore. The RD Servers has code, its old but its there, that was supposed to handle techweb point generation. If they overheat, they would to slow down point generation till someone turned the freezer back on. Alas all of that code was skipped over in favor of a hard constant in the research subsystem. Before this PR, as long as one server was somewhere in the world, point generation would continue.

So here is what I did:

    I upgraded the RD server console to tgui2 (tgui 3 tested and working). Its better than it was before but until a refactor of the logging code on the RD Console is done, not a lot of reasons to do more.

    The servers run fine when the temp is below 0C, and stop generating points once they reach 20C. The server lights show the status as well as the console.

    Once you overheat the servers, they will start sparkling and won't return to normal till the temperature gets below 0C again.

    Yes, you can build more servers for more point generation, however, the scaling factor kills any reason having more than 3 at one time. Not to mention you need to upgrade the server room the more servers you add as that lonely T1 freezer just barley keeps two servers down in the green.

    I implemented the server heat effect as a component, so now you can make other machines or atoms passively heat up a room! Will probably come back to it to implement endothermic mode, like for ice blocks or such.

This PR currently enables the servers on default, however, you can disable this entire system by changing SSresearch.multiserver_calculation to FALSE.
```

## Why It's Good For The Game
From the original PR:
```
These things have had their own little room for more than 6 years and has sit there while people paint the walls with monkey blood. It makes sense, lore wise, that there would also be a place that all these research points are stored in rather it just generated in the eather. Maybe down the line we could have the points and tech physically stored in the servers so if some idiot blows up the room, someone gets fired for not making disk backups.

Well holding off on that till techweb gets cleaned up.

Currently there shouldn't be any map issues. I tested it with Box, Meta, Delta, and even Kilo and they all operate the same. Have had it run 4 hours in the background on the chance I miscalculated the heat values but they still chugging along.
```

## Changelog
:cl:WarlockD
add: Research Servers now actually generate points, and are sensitive to heat.
/:cl:

### CODE REVIEW

**Comment by SomeAngryMiner** [comment]
blessed.

**Comment by Coolsurf6** [comment]
Could you maybe upgrade the servers as well or upgrade the cooling to increase point generation? That might be spicy.

**Comment by St0rmC4st3r** [comment]
Nice! Waiting for a PR that will make telecomms depending on coolers!

**Comment by WarlockD** [comment]
Yo thanks for putting it back.  I just shrugged it off once it was closed with no discussion so didn't think anyone cared:)  My handle on discord is Erodros and same on Byond.

There are going to be a few problems though.  There needs to be documentation, the component needs some work or just delete the component and use the machine process to emit the heat.  I was about half way done with this till it was closed and I just deleted the fork to work on something else.  Good luck:)

**Comment by MCterra10** [comment]
pretty sure bee doesn't care about dmdoc. I'll look into the component stuff on the weekend, probably. Might just be easier to stick it in the process.

File: code/datums/components/thermo.dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+/datum/component/thermo/Initialize(_targetTemperature = T20C, _heatingPower = 10000, _efficiency = 0.25)
+	targetTemperature = _targetTemperature
+	heatingPower = _heatingPower
+	efficiency = _efficiency
+	START_PROCESSING(SSobj, src)
+
+/datum/component/thermo/UnregisterFromParent()
+	STOP_PROCESSING(SSobj, src)
+
+
```
Comment (Kmc2000) [inline code/datums/components/thermo.dm]: whitespace

**Review by Kmc2000 — APPROVED** [review]

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
2 things I noticed, not sure if they are intentional or not
1. Servers that are turned off reduce the amount of research being made, meaning golems on lavaland can build hundreds of servers and despite having no power the amount of research being generated will fall to almost nothing.
2. Hot servers reduce the amount of research being made.

**Comment by MCterra10** [comment]
the second is most certainly intentional.
can you be a bit more descriptive about the first @PowerfulBacon ?

**Comment by MCterra10** [comment]
will rebuild tgui to fix the conflicts soon

File: code/datums/components/thermo.dm (hunk 1 above)
Comment (MCterra10) [reply code/datums/components/thermo.dm]: pretty trivial review considering how many files have odd numbers of newlines, but I'll make a commit for it regardless.

**Comment by PowerfulBacon** [comment]
> 
> 
> the second is most certainly intentional.
> can you be a bit more descriptive about the first @PowerfulBacon ?

Build a server on lavaland with no power, the global amount of research being produced decreases, despite having no power, being turned off and not even being on the station.

**Comment by WarlockD** [comment]
Humm.  Yea.  I didn't intend that.  Basically I was using that sqrt formula to calculate the diminishing returns.  So after the 4th server, your only getting like 100 extra points a min or such.  I see the problem though, I didn't want to touch techweb much, but it counts the servers and does the math after it runs mine().  So when mine() returns 0 on unpowered servers it screws with the count.  It needs to check if the server is running instead of just using the len count.

Still, pretty funny for a bunch of golems to screw the station that way hah.

**Comment by WarlockD** [comment]
Oh side note this was my first pr, so things like storing the component as a var and other issues, I am sorry:P

**Comment by MCterra10** [comment]
that should fix the issue with wrong research amounts. the research subsystem no longer includes inactive servers in its research coefficient calculations

**Comment by MCterra10** [comment]
I'll work on moving stuff to the machine process and then do some tests

**Comment by MCterra10** [comment]
oh you're kidding me, it compiled locally one sec

**Comment by MCterra10** [comment]
Tested. This is working exactly as intended now.
Unpowered or non-operating RND servers no longer lower the per-server coefficient.
All other features working as expected.
Perhaps I'll re-add the component another time

**Comment by MCterra10** [comment]
rip your new label bot doesn't support it

**Comment by PowerfulBacon** [comment]
I'll have a look through this again

**Review by PowerfulBacon — APPROVED** [review]

**Comment by ike709** [comment]
Seems fine.

**Comment by MCterra10** [comment]
winrar

**Pull request merged**